File: src/main/java/com/offz/spigot/custommobs/Mobs/MobDrop.java
package com.offz.spigot.custommobs.Mobs;

import org.bukkit.Material;
import org.bukkit.configuration.serialization.ConfigurationSerializable;
import org.bukkit.inventory.ItemStack;

import java.util.Map;

public class MobDrop implements ConfigurationSerializable {
    private ItemStack item;
    private double dropChance;
    private int minAmount;
    private int maxAmount;

    //TODO make work with ItemStack
    public MobDrop(Material material, int amount) {
        this(material, amount, amount, 1);
    }

    public MobDrop(Material material, int amount, double dropChance) {
        this(material, amount, amount, dropChance);
    }

    public MobDrop(Material material, int minAmount, int maxAmount) {
        this(material, minAmount, maxAmount, 1);
    }

    public MobDrop(Material material, int minAmount, int maxAmount, double dropChance) {
        this(new ItemStack(material), minAmount, maxAmount, dropChance);
    }

    public MobDrop(ItemStack item, int minAmount, int maxAmount, double dropChance) {
        this.item = item;
        this.dropChance = dropChance;
        this.minAmount = minAmount;
        this.maxAmount = maxAmount;
    }

    /**
     * Required method for configuration serialization
     *
     * @param args map to deserialize
     * @return deserialized item stack
     * @see ConfigurationSerializable
     */
    public static MobDrop deserialize(Map<String, Object> args) {
        ItemStack item;
        int minAmount;
        int maxAmount;
        double dropChance = 1;

        if (args.containsKey("item"))
            item = ((ItemStack) args.get("item"));
        else if (args.containsKey("material"))
            item = new ItemStack(Material.getMaterial((String) args.get("material")));
        else
            return null;

        if (args.containsKey("drop_chance"))
            dropChance = (Double) args.get("drop_chance");

        if (args.containsKey("amount")) {
            int amount = (Integer) args.get("amount");
            return new MobDrop(item, amount, amount, dropChance);
        }

        if (!(args.containsKey("min_amount") && args.containsKey("max_amount")))
            return null;

        minAmount = (Integer) args.get("min_amount");
        maxAmount = (Integer) args.get("max_amount");

        return new MobDrop(item, minAmount, maxAmount, dropChance);
    }

    public ItemStack chooseDrop() {
        if (Math.random() < dropChance) {
            ItemStack drop = item.clone();
            drop.setAmount((int) ((Math.random() * (maxAmount - minAmount)) + minAmount));
            return drop;
        }
        return null;
    }

    @Override
    public Map<String, Object> serialize() {
        return null;
    }
}

File: src/main/java/com/mineinabyss/mobzy/ecs/components/death/MobDrop.kt
package com.mineinabyss.mobzy.ecs.components.death

import com.mineinabyss.idofront.messaging.color
import com.mineinabyss.idofront.recipes.addCookingRecipes

import com.mineinabyss.idofront.serialization.SerializableItemStack
import com.mineinabyss.mobzy.mobzy
import kotlinx.serialization.Serializable
import org.bukkit.ChatColor
import org.bukkit.inventory.ItemStack
import kotlin.math.roundToInt
import kotlin.random.Random

/**
 * A serializable class for defining an item a mob can drop. It acts as a sort of factory for [ItemStack]s that may
 * define rarity, how many items to spawn, etc...
 *
 * Also allows defining a cooked version of the item that will automatically be registered when the class is created.
 *
 * @param item The item to be spawned.
 * @param cooked The cooked version of this item.
 * @param cookExp The amount of exp to drop when this item is cooked.
 * @param cookTime How long it takes for this item to cook
 * @param minAmount The minimum stack size of this item.
 * @param maxAmount The maximum stack size of this item.
 * @param dropChance A chance from 0 to 1 for this item to be dropped.
 *
 */
@Serializable
data class MobDrop(
    val item: SerializableItemStack,
    val cooked: SerializableItemStack? = null,
    val cookExp: Float = 0f,
    val cookTime: Int = 200,
    val minAmount: Int = 1,
    val maxAmount: Int = 1,
    val dropChance: Double = 1.0
) {
    init {
        if (cooked != null) {
            val cookedItem = cooked.toItemStack()
            //TODO pretty hacky way of adding recipes. Won't reload them from config if they've changed.
            try {
                addCookingRecipes(
                    ChatColor.stripColor(cookedItem.itemMeta!!.displayName.color('§').replace(' ', '_'))!!,
                    item.toItemStack(),
                    cookedItem,
                    cookExp,
                    cookTime,
                    mobzy
                )
            } catch (e: IllegalStateException) {
            }
        }
    }

    /** @return The amount of items to be dropped, or null if the drop does not succeed */
    // TODO I'd like to use exactly what Minecraft's existing system is, but I can't seem to find a way to reuse that.
    fun chooseDrop(lootingLevel: Int, fire: Boolean): ItemStack? {
        val lootingPercent = lootingLevel / 100.0

        val lootingMaxAmount: Int =
            if (dropChance >= 0.5)
                (maxAmount + lootingLevel * Random.nextDouble()).roundToInt()
            else maxAmount

        val lootingDropChance =
            if (dropChance >= 0.10)
                dropChance * (10.0 + lootingLevel) / 10.0
            else dropChance + lootingPercent

        return if (Random.nextDouble() < lootingDropChance) {
            val drop =
                if (fire && cooked != null)
                    cooked.toItemStack()
                else item.toItemStack()
            drop.amount =
                if (lootingMaxAmount <= minAmount) minAmount
                else Random.nextInt(minAmount, lootingMaxAmount)
            drop
        } else null
    }
}
